propolis-server/src/api.rs

```diff
@@ -0,0 +1,189 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstancePathParams {
+    pub instance_id: Uuid,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceEnsureRequest {
+    pub properties: InstanceProperties,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceEnsureResponse {}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceGetResponse {
+    pub instance: Instance,
+}
+
+/// Requested state of an Instance.
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceStateChange {
+    pub state: InstanceStateRequested,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub enum InstanceStateRequested {
+    Run,
+    Stop,
+    Reboot,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceSerialRequest {
+    pub bytes: Vec<u8>,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceSerialResponse {
+    pub bytes: Vec<u8>,
+}
+
+/// Current state of an Instance.
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub enum InstanceState {
+    Creating,
+    Starting,
+    Running,
+    Stopping,
+    Stopped,
+    Repairing,
+    Failed,
+    Destroyed,
+}
+
+#[derive(Clone, Deserialize, Serialize, JsonSchema)]
+pub struct InstanceProperties {
+    pub generation_id: u64,
+
+    /// Unique identifier for this Instance.
+    pub id: Uuid,
+    /// Human-readable name of the Instance.
+    pub name: String,
+    /// Free-form text description of an Instance.
+    pub description: String,
+
+    /// ID of the image used to initialize this Instance.
+    pub image_id: Uuid,
+    /// ID of the bootrom used to initialize this Instance.
+    pub bootrom_id: Uuid,
+    /// Size of memory allocated to the Instance, in MB.
```

Comment:
You're totally right - I was too fast and loose here, I'll tighten up the referenced allocations (MiB > MB!).

I do think there is an advantage to using MiB over bytes - we don't really want to enable byte-granularity allocations, so by calling this MiB, it's harder to misuse the API (sure, we can round up if someone asks for a fraction of a MiB, but it would be nice to just prevent that weird request in the first place).

On the other hand, consistency is nice for using "individual bytes" throughout. I could be persuaded either way,  but I'll proooobably stick with MiB in the meantime?

Reply:
Yeah, MiB sounds like a good approach here. The `ApiByteCount` type supports constructions from mebibytes, among other things, which we could leverage to make sure the conversion from bytes in the public API is an appropriate multiple of an MiB.